File: Provenance/Categories/UIView+FrameAdditions.m
#import "UIView+FrameAdditions.h"


@implementation UIView (UIView_FrameAdditions)

- (void)setOrigin:(CGPoint)origin
{
	CGRect frame = [self frame];
	frame.origin = origin;
	[self setFrame:frame];
}

- (void)setOriginX:(CGFloat)originX
{
	CGRect frame = [self frame];
	frame.origin.x = originX;
	[self setFrame:frame];
}

- (void)setOriginY:(CGFloat)originY
{
	CGRect frame = [self frame];
	frame.origin.y = originY;
	[self setFrame:frame];
}

- (void)setSize:(CGSize)size
{
	CGRect frame = [self frame];
	frame.size = size;
	[self setFrame:frame];
}

- (void)setHeight:(CGFloat)height
{
	CGRect frame = [self frame];
	frame.size.height = height;
	[self setFrame:frame];
}

- (void)setWidth:(CGFloat)width
{
	CGRect frame = [self frame];
	frame.size.width = width;
	[self setFrame:frame];
}

@end

// UIAppearance+Swift.m
#if TARGET_OS_IOS
@implementation UIView (UIViewAppearance_Swift)
+ (instancetype)my_appearanceWhenContainedIn:(Class<UIAppearanceContainer>)containerClass {
	return [self appearanceWhenContainedIn:containerClass, nil];
}
@end
#endif

File: Provenance/Categories/UIView+FrameAdditions.swift
import UIKit

@objc
public protocol HitAreaEnlarger: class {
    var hitAreaInset: UIEdgeInsets { get set }
}

public extension HitAreaEnlarger where Self: UIButton {
    var touchAreaEdgeInsets: UIEdgeInsets {
        get {
            if let value = objc_getAssociatedObject(self, &hitAreaInset) as? NSValue {
                var edgeInsets: UIEdgeInsets = .zero
                value.getValue(&edgeInsets)
                return edgeInsets
            } else {
                return .zero
            }
        }
        set(newValue) {
            var newValueCopy = newValue
            let objCType = NSValue(uiEdgeInsets: .zero).objCType
            let value = NSValue(&newValueCopy, withObjCType: objCType)
            objc_setAssociatedObject(self, &hitAreaInset, value, .OBJC_ASSOCIATION_RETAIN)
        }
    }

    func point(inside point: CGPoint, with event: UIEvent?) -> Bool {
        if touchAreaEdgeInsets == .zero || !isEnabled || isHidden {
            return (self as UIControl).point(inside: point, with: event)
        }

        let relativeFrame = bounds
        let hitFrame = relativeFrame.inset(by: touchAreaEdgeInsets)

        return hitFrame.contains(point)
    }
}

public extension UIView {
    func setOrigin(_ origin: CGPoint) {
        var frame: CGRect = self.frame
        frame.origin = origin
        self.frame = frame
    }

    func setOriginX(_ originX: CGFloat) {
        var frame: CGRect = self.frame
        frame.origin.x = originX
        self.frame = frame
    }

    func setOriginY(_ originY: CGFloat) {
        var frame: CGRect = self.frame
        frame.origin.y = originY
        self.frame = frame
    }

    func setSize(_ size: CGSize) {
        var frame: CGRect = self.frame
        frame.size = size
        self.frame = frame
    }

    func setHeight(_ height: CGFloat) {
        var frame: CGRect = self.frame
        frame.size.height = height
        self.frame = frame
    }

    func setWidth(_ width: CGFloat) {
        var frame: CGRect = self.frame
        frame.size.width = width
        self.frame = frame
    }
}
